### PR TITLE
Day & Night Cycle Rule

### PR DESCRIPTION
**Pull request by ThebestkillerTBK**
The game will be able to cycle the time in custom mode. This function is similar to the campaign planet light. [https://github.com/Anuken/Mindustry-Suggestions/issues/2085](https://github.com/Anuken/Mindustry-Suggestions/issues/2085)
Test map: [map_138.zip](https://github.com/Anuken/Mindustry/files/6554532/map_138.zip)

### CODE REVIEW

**Comment by ThebestkillerTBK** [comment]
> Huh, i actually remember making a suggestion exactly adding this

Which issue?

**Comment by Anuken** [comment]
[This violates several of the contribution guidelines.](https://github.com/Anuken/Mindustry/blob/master/CONTRIBUTING.md)

**Comment by MEEPofFaith** [comment]
s t y l e

File: core/src/mindustry/game/Rules.java [hunk 1]
```diff
@@ -90,6 +90,12 @@
     public boolean lighting = false;
     /** Whether enemy lighting is visible.
      * If lighting is enabled and this is false, a fog-of-war effect is partially achieved. */
+    public boolean dayNightCycle = false;
+    /** Whether Day & Night Cycle is enabled. */
+    public float dayNightCycleSpeed = 1.5f;
+    /** The light changing speed.*/
+    public boolean dayNightCycleDark = false;
+    /** Whether the light getting darker or brighter.*/
     public boolean enemyLights = true;
```
Comment (MEEPofFaith) [inline core/src/mindustry/game/Rules.java:99]:
```suggestion
    public boolean lighting = false;
    /** Whether enemy lighting is visible.
     * If lighting is enabled and this is false, a fog-of-war effect is partially achieved. */
    public boolean enemyLights = true;
    /** Whether Day & Night Cycle is enabled. */
    public boolean dayNightCycle = false;
    /** The light changing speed.*/
    public float dayNightCycleSpeed = 1.5f;
    /** Whether the light getting darker or brighter.*/
    public boolean dayNightCycleDark = false;
```

Do you even know how these comments work?

File: core/src/mindustry/core/Logic.java [hunk 2]
```diff
@@ -267,6 +270,31 @@ private void updateWeather(){
...
+                state.rules.dayNightCycleDark = true;
+            }
+
+            if (state.rules.dayNightCycleDark){
+                state.rules.ambientLight.a = state.rules.ambientLight.a + realSpeed;
+            } else {
+                state.rules.ambientLight.a = state.rules.ambientLight.a - realSpeed;
+            }
+            
+        }
```
Comment (MEEPofFaith) [inline core/src/mindustry/core/Logic.java:295]:
```suggestion
        if(state.rules.dayNightCycle){
            if(state.rules.ambientLight.a >= 1){
                state.rules.ambientLight.a = 0;
            }
            if(state.rules.ambientLight.a + realSpeed > 1 || state.rules.ambientLight.a > darkLight){
                state.rules.ambientLight.a = darkLight;
                state.rules.dayNightCycleDark = false;
            }
            if(state.rules.ambientLight.a - realSpeed < 0){
                state.rules.ambientLight.a = 0;
                state.rules.dayNightCycleDark = true;
            }

            if(state.rules.dayNightCycleDark){
                state.rules.ambientLight.a = state.rules.ambientLight.a + realSpeed;
            }else{
                state.rules.ambientLight.a = state.rules.ambientLight.a - realSpeed;
            }
        }
```

F o r m a t t I n g

File: core/src/mindustry/core/Logic.java [hunk 3]
```diff
@@ -383,17 +411,30 @@ public void update(){
...
                     for(TeamData data : state.teams.getActive()){
                         if(data.hasAI()){
                             data.ai.update();
                         }
                     }
                 }
 
+                if (net.server() && state.rules.dayNightCycle){
+                    Call.setRules(state.rules);
+                }
```
Comment (MEEPofFaith) [inline core/src/mindustry/core/Logic.java:436]:
```suggestion
                    if(state.hasSector()){
                        state.rules.dayNightCycle = false;
                    }
                    if(state.rules.dayNightCycle && state.rules.lighting == false){
                        state.rules.lighting = true;
                        state.rules.ambientLight = new Color(0.01f, 0.01f, 0.04f, 0.1f);
                    };
                    dayNightCycleUpdate();

                    for(TeamData data : state.teams.getActive()){
                        if(data.hasAI()){
                            data.ai.update();
                        }
                    }
                }

                if(net.server() && state.rules.dayNightCycle){
                    Call.setRules(state.rules);
                }
```

F o r m a t t i n g

**Comment by Anuken** [comment]
Stale, can't be merged in current state - spams TCP packets with setRules, creates new Color instances in main loop, etc.

**Comment by vicfic18** [comment]
Will this ever be implemented?

**Comment by Anuken** [comment]
Not unless someone creates a better implementation.

**Pull request closed without merging**